Add internal IRT implementation of execv()/execve()

alyptik opened the pull request:
Add `execv()` and `execve()` system call implementations to the internal runtime.

Implementation Details:
- Copy the argument and environment vectors from the "parent" NaClApp.
- Duplicate its `cage_id` along with any other relevant internal state.
- Leverage the existing fork() implementation to emulate replacement of the entire process image.

alyptik committed:
linting: clean up NaClPatchAddr() function

Signed-off-by: Joey Pabalinas <joeypabalinas@gmail.com>

alyptik committed:
sel_ldr: add additional NaClApp structure members

Add members to hold process metadata and rename existing members to
better reflect usage.

Signed-off-by: Joey Pabalinas <joeypabalinas@gmail.com>

alyptik committed:
sel_ldr: add mutex guards against spurious thread spawns

Calling pthread_create() on glibc 2.28+ spawns extra threads (underlying
cause is still unknown) which ultimately results in the spawning of more
than one forked child.

Add mutexes to stop these spurious thread spawns and only spawn one
child per fork() call.

Signed-off-by: Joey Pabalinas <joeypabalinas@gmail.com>

alyptik committed:
getppid: add internal IRT implementation of getppid()

Signed-off-by: Joey Pabalinas <joeypabalinas@gmail.com>

alyptik committed:
dup: correctly duplicate file descriptors in dup()/dup2()/dup3()

Signed-off-by: Joey Pabalinas <joeypabalinas@gmail.com>

alyptik committed:
sel_main: update struct member names during NaCl initialization phase

Signed-off-by: Joey Pabalinas <joeypabalinas@gmail.com>

alyptik committed:
pipe: update IO-related system calls to support pipe() emulation

Signed-off-by: Joey Pabalinas <joeypabalinas@gmail.com>

alyptik committed:
syscall_handlers: add new system calls and update prototypes

Signed-off-by: Joey Pabalinas <joeypabalinas@gmail.com>

alyptik committed:
fork: add supporting system calls for fork() and cleanup existing implementations

Signed-off-by: Joey Pabalinas <joeypabalinas@gmail.com>

alyptik committed:
fork: update fork() implementation to support execve()

Update NaClApp member names and simplify implementation.

Signed-off-by: Joey Pabalinas <joeypabalinas@gmail.com>

alyptik committed:
execve: add internal runtime implementation of execv()/execve()

Signed-off-by: Joey Pabalinas <joeypabalinas@gmail.com>

JustinCappos commented on src/trusted/service_runtime/nacl_globals.h:
```diff
@@ -24,6 +24,7 @@ EXTERN_C_BEGIN
 #define PROT_RW (NACL_ABI_PROT_READ | NACL_ABI_PROT_WRITE)
 #define PROT_RX (NACL_ABI_PROT_READ | NACL_ABI_PROT_EXEC)
 #define F_ANON_PRIV (NACL_ABI_MAP_PRIVATE | NACL_ABI_MAP_ANONYMOUS)
+#define SHM_SIZE (1u << 13)
```
why?  Also, will this always be a constant?

alyptik replied:
Whoops, that was leftover code from something else I had planned but decided wasn't worth pursuing! Good catch, removing.

JustinCappos commented on src/trusted/service_runtime/sel_ldr_standard.c:
```diff
@@ -727,11 +725,11 @@ int NaClCreateMainThread(struct NaClApp     *nap,
    * code does not look like string data....
    */
-  for (i = 0; i < argc && argv; ++i) {
+  for (i = 0; i < argc && argv && argv[i]; ++i) {
```
Why ++i instead of i++?

alyptik replied:
The rest of the `for()` loops in this file use `++i`, I figured it'd be good to stay consistent.

moyix commented on src/trusted/service_runtime/nacl_syscall_common.c:
```diff
@@ -3965,38 +3965,32 @@ int32_t NaClSysPipe2(struct NaClAppThread *natp, uint32_t *pipedes, int flags) {
     NaClLog(1, "%s\n", "[NaClSysFork] forking program failed!");
-    ret = -NACL_ABI_ENOMEM;
-    goto out;
+    ret = -NACL_ABI_ENOEXEC;
```
Why the change to ENOEXEC here? I don't think that's a standard error for `fork` (it is for exec, e.g. in case the thing you're trying to exec isn't a properly formatted executable).

alyptik replied:
After checking the manpage and POSIX standard, it seems you are right and `ENOEXEC` isn't even a valid `errno`. Changing it back to `ENOMEM`.

moyix commented on src/trusted/service_runtime/nacl_syscall_common.c:
```diff
@@ -456,62 +456,88 @@ int32_t NaClSysNameService(struct NaClAppThread *natp,
+    goto out;
+  }
+  new_desc = NaClSetAvail(nap, old_nd);
+  fd_cage_table[nap->cage_id][newfd] = new_desc;
```
Does `newfd` need to first be closed if it's open (per the `dup2` man page)? Or is that done by `NaClSetAvail` somehow?

alyptik replied:
Good catch, I completely forget to close `newfd`. Fixing.

moyix commented on src/trusted/service_runtime/nacl_syscall_common.c:
```diff
@@ -456,62 +456,88 @@ int32_t NaClSysNameService(struct NaClAppThread *natp,
 int32_t NaClSysDup3(struct NaClAppThread  *natp,
                     int                   oldfd,
                     int                   newfd,
                     int                   flags) {
```
As a general point, can `dup` and `dup2` be implemented as calls to `dup3` with the appropriate flags?

alyptik replied:
Hm, well `dup()` not really, since it returns an arbitrary file descriptor, but I think it may be possible with `dup2()`/`dup3`. I'll see if I can rework `dup3()` to just call `dup2()`.

alyptik replied:
Duplicated code removed.

moyix commented on src/trusted/service_runtime/nacl_syscall_common.c:
```diff
@@ -2792,7 +2790,7 @@ int32_t NaClSysImcRecvmsg(struct NaClAppThread         *natp,
 
   recv_hdr.ndescv = new_desc;
-  recv_hdr.ndesc_length = NACL_ARRAY_SIZE(new_desc);
+  recv_hdr.ndesc_length = sizeof new_desc / sizeof *new_desc;
```
Would prefer to always use parentheses with sizeof – makes things a bit easier to read and removes ambiguity.

alyptik replied:
Sure, I can do that.

moyix commented on src/trusted/service_runtime/nacl_syscall_common.c:
```diff
@@ -2881,7 +2879,7 @@ int32_t NaClSysImcRecvmsg(struct NaClAppThread         *natp,
   /* copy out updated desc count, flags */
   if (retval < 0) {
-    for (i = 0; i < NACL_ARRAY_SIZE(new_desc); ++i) {
+    for (i = 0; i < sizeof new_desc / sizeof *new_desc; ++i) {
```
Parentheses again

alyptik replied:
Will fix.

alyptik committed:
dup2: close newfd before making it a copy of oldfd

Signed-off-by: Joey Pabalinas <joeypabalinas@gmail.com>

alyptik committed:
linting: add parentheses around sizeof operands

Signed-off-by: Joey Pabalinas <joeypabalinas@gmail.com>

alyptik committed:
close: mark closed file descriptors as invalid

Signed-off-by: Joey Pabalinas <joeypabalinas@gmail.com>

alyptik committed:
linting: change `++i` in for loop controlling expressions to `i++`

The postfix `i++` expression is more conventional and thus
preferred in for loop controlling expressions as the two
have equivalent semantics.

Signed-off-by: Joey Pabalinas <joeypabalinas@gmail.com>

alyptik committed:
dup3: call dup2 directly instead of duplicating code

Signed-off-by: Joey Pabalinas <joeypabalinas@gmail.com>

alyptik commented:
Thank you for the quick review, merged.
